tests/qnodes/test_qnode_passthru.py

```diff
@@ -0,0 +1,186 @@
+def tensornet_tf_device():
+    return qml.device('default.tensor.tf', wires=2)
+
+
+@pytest.mark.skipif(tf is None, reason="TensorFlow 2.0 not found.")
+class TestPassthruTF:
+    """Test that TF objects can be successfully passed through to a TF simulator device, and back to user."""
+
+    # real data type used by the default.tensor.tf plugin (TensorFlow is strict about types)
+    DTYPE = tf.float64 if tf else None
+
+    def test_arraylike_args(self, tensornet_tf_device, tol):
+        """Tests that PassthruQNode can use array-like TF objects as positional arguments."""
+
+        def circuit(x):
+            qml.RX(x[0], wires=[0])
+            qml.RX(2*x[1], wires=[1])
+            return qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))
+
+        node = PassthruQNode(circuit, tensornet_tf_device)
+        x = tf.Variable([1.1, 1.4], dtype=self.DTYPE)
+        res = node(x)
+        assert isinstance(res, tf.Tensor)
+        assert res.shape == (2,)
+        assert res.dtype == self.DTYPE
+
+    def test_arraylike_keywordargs(self, tensornet_tf_device, tol):
+        """Tests that qnodes use array-like TF objects as keyword-only arguments."""
+
+        def circuit(*, x=None):
+            qml.RX(x[0], wires=[0])
+            qml.RX(2*x[1], wires=[1])
+            return qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))
+
+        node = PassthruQNode(circuit, tensornet_tf_device)
+        x = tf.Variable([1.1, 1.4], dtype=self.DTYPE)
+        res = node(x=x)
+        assert isinstance(res, tf.Tensor)
+        assert res.shape == (2,)
+        assert res.dtype == self.DTYPE
+
+    def test_tensor_operations(self, tensornet_tf_device, tol):
+        """Tests the evaluation of a PassthruQNode involving algebraic operations between tensor parameters,
+        and TF functions acting on them."""
+
+        def circuit(phi, theta):
+            x = phi * theta
+            qml.RX(x[0], wires=0)
+            qml.RY(tf.cos(x[1]), wires=1)
+            qml.CNOT(wires=[0, 1])
+            return qml.expval(qml.PauliZ(0)), qml.expval(qml.Hadamard(1))
+
+        node = PassthruQNode(circuit, tensornet_tf_device)
+
+        phi = tf.Variable(np.array([0.7, -1.2]), dtype=self.DTYPE)
+        theta = tf.Variable(1.7, dtype=self.DTYPE)
+        res = node(phi, theta)
+        assert isinstance(res, tf.Tensor)
+        assert res.shape == (2,)
+        assert res.dtype == self.DTYPE
+
+    def test_evaluate(self, tensornet_tf_device, tol):
+        """Tests the evaluation of a PassthruQNode."""
+
+        def circuit(x):
+            qml.RX(x[0], wires=0)
+            qml.RY(x[1], wires=1)
+            qml.CNOT(wires=[0, 1])
+            qml.RX(x[2], wires=1)
+            return qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))
+
+        x = np.array([-1.4, 0.2, 2.1])
+        x_tf = tf.Variable(x, dtype=self.DTYPE)
+
+        node = PassthruQNode(circuit, tensornet_tf_device)
+        res = node(x_tf)
+
+        # compare to a reference node
```

Comment:
This circuit is simple enough to work out the answer by hand. Might be good to add that instead of using reference?

Reply:
Ehh, maybe using Mathematica or Sympy, it has three rotations in it :)

Reply:
Done.